pkg/controller/registry/resolver/cache.go

```diff
@@ -0,0 +1,287 @@
+		list, err := rcp.c.OperatorsV1alpha1().CatalogSources(namespace).List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			rcp.logger.Errorf("failed to list catalogsources in %s: %s", namespace, err.Error())
+			continue
+		}
+		for _, source := range list.Items {
+			if source.Status.RegistryServiceStatus == nil {
+				continue
+			}
+			key := CatalogKey{
+				Namespace: source.Namespace,
+				Name:      source.Name,
+			}
+			result[key] = &LazyClient{address: source.Address()}
+		}
+	}
+	return result
+}
+
+type CatalogDependencyCache interface {
+	GetCSVNameFromCatalog(csvName string, catalog CatalogKey) (Operator, error)
+	GetCSVNameFromAllCatalogs(csvName string) ([]Operator, error)
+	GetPackageFromAllCatalogs(pkg string) ([]Operator, error)
+	GetPackageVersionFromAllCatalogs(pkg string, version semver.Version) ([]Operator, error)
+}
+
+type OperatorCache struct {
+	logger    logrus.FieldLogger
+	rcp       RegistryClientProvider
+	snapshots map[CatalogKey]*CatalogSnapshot
+	ttl       time.Duration
+	sem       chan struct{}
+	m         sync.RWMutex
+}
+
+func NewOperatorCache(rcp RegistryClientProvider) *OperatorCache {
+	return &OperatorCache{
+		logger:    logrus.New(),
+		rcp:       rcp,
+		snapshots: make(map[CatalogKey]*CatalogSnapshot),
+		ttl:       5 * time.Minute,
+		sem:       make(chan struct{}, 4),
+	}
+}
+
+type NamespacedOperatorCache struct {
+	snapshots map[CatalogKey]*CatalogSnapshot
+}
+
+var _ CatalogDependencyCache = &NamespacedOperatorCache{}
+
+func (c *OperatorCache) Namespaced(namespaces ...string) *NamespacedOperatorCache {
+	now := time.Now()
+	clients := c.rcp.ClientsForNamespaces(namespaces...)
+
+	result := NamespacedOperatorCache{
+		snapshots: make(map[CatalogKey]*CatalogSnapshot),
+	}
+
+	var misses []CatalogKey
+	c.m.RLock()
+	for key := range clients {
+		if snapshot, ok := c.snapshots[key]; ok && !snapshot.Expired(now) {
+			result.snapshots[key] = snapshot
+		} else {
+			misses = append(misses, key)
+		}
+	}
+	c.m.RUnlock()
+
+	if len(misses) == 0 {
+		return &result
+	}
+
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	// Take the opportunity to clear expired snapshots while holding the lock.
+	var expired []CatalogKey
+	for key, snapshot := range c.snapshots {
+		if snapshot.Expired(now) {
+			expired = append(expired, key)
+		}
+	}
+	for _, key := range expired {
+		delete(c.snapshots, key)
+	}
+
+	// Check for any snapshots that were populated while waiting to acquire the lock.
+	var found int
+	for i := range misses {
+		if snapshot, ok := c.snapshots[misses[i]]; ok && !snapshot.Expired(now) {
+			result.snapshots[misses[i]] = snapshot
+			misses[found], misses[i] = misses[i], misses[found]
+			found++
+		}
+	}
+	misses = misses[found:]
+
+	for _, miss := range misses {
+		s := CatalogSnapshot{
+			logger: c.logger.WithField("catalog", miss),
+			key:    miss,
+			expiry: now.Add(c.ttl),
+		}
+		s.m.Lock()
+		c.snapshots[miss] = &s
+		result.snapshots[miss] = &s
+		go c.populate(&s, clients[miss])
```

Comment:
self: check that we're tracking in-flight populates

Reply:
Thanks, I had forgotten to add a timeout to the RPC in populate and cancellation on snapshot eviction. Updating that now.